### ISSUE TITLE
Crash after release of shared image (regression)

### ISSUE THREAD

**Issue by wcout** [fltk/fltk#887]
**Describe the bug**
There seems to be a regression with this commit: https://github.com/fltk/fltk/commit/c10183379fdc0497f9275385d23d174c0f2d4461

**To Reproduce**
See attached program [crash_shimg.cxx.txt](https://github.com/fltk/fltk/files/13920286/crash_shimg.cxx.txt)

1. Run the program
2. Select an image file e.g. some .PNG file (I was using images from the documention/src folder), which i had generated
3. The image is displayed in a window
4. close the window
5. The file chooser should be displayed again, but here the program crashes (not before the above commit)
6. Repeat some times if the crash does not occur immediately

**Expected behavior**
The program should not crash

**FLTK Version**
1.4.0 git current

**Operating System / Platform:**
Ubuntu 20.04

**Linux/Unix Runtime, if applicable:**
 - X11 and Wayland

**Comment by MatthiasWM** [comment]
Thanks. Those are my changes. I will look at that.

**Comment by MatthiasWM** [comment]
This should be fixed now. Please let me know if this solves your issue, so we can close this report.

**Comment by wcout** [comment]
Well, there is no crash now.
But I don't understand why the shared images refcount after the initial ::get() in my program is already at 2 now.

**Comment by MatthiasWM** [comment]
Sigh, you are right.

**Comment by MatthiasWM** [comment]
Ok, this should be fixed now. The file chooser was keeping a reference to the preview image, even when hidden. And because the preview image is resized, it keeps another reference to the original image. Add your reference from your app, and you have two references to the original and one to the resized image. The *scaled* image does not generate another reference, but that is a whole other story.

BTW: if you call `b->bind_image();` instead of `b->image();`, FLTK will manage the shared image for you and release the reference when you delete the widget or assign a different image.

**Comment by MatthiasWM** [comment]
https://fltk.gitlab.io/fltk/classFl__Widget.html#a3b39a8940aeb832156ece4945e8beb88

**Comment by wcout** [comment]
Phew, so many changes for that. I can't verify them, just see that it seems to work properly now.

Offtopic: I notice, that current FLTK 1.3 branch also has some problem with this situation. It does not crash, but the file chooser shows a big 'X' (broken image sign) after re-opening. This was not so with FLTK 1.3.8.

**Comment by Albrecht-S** [comment]
> Offtopic: I notice, that current FLTK 1.3 branch also has some problem with this situation. .... This was not so with FLTK 1.3.8.

@wcout Thank you very much for your tests and for reporting issues. This is not offtopic here because the main part of the related changes have also been backported to 1.3.9 which obviously has some issues.

@MatthiasWM Please take a look at branch-1.3, there we have these related commits:
```
783d50e6b 2023-12-06 16:27:10 +0100 Matthias Melcher   - #188: Fixes reference count for Fl_Shared_Image
f655a4d93 2023-12-06 14:29:32 +0100 Matthias Melcher   - #188: Fixes search for Fl_SHared_Image original
```

If this turned out to be an issue that **_can cause crashes_** etc. (illegal memory references) we might need to release another patch release (1.3.10). Thanks in advance.

**Comment by MatthiasWM** [comment]
@Albrecht-S Yes, sorry for replacing one bug with another in 1.3 . I will try to figure this one out.

**Comment by wcout** [comment]
Sorry, but b2dc8217a17f0e89b303461743965f2b3699a570 does not fix the issue in branch 1.3. The test program crashes or shows the big 'X'.

**Comment by MatthiasWM** [comment]
See #188.

**Comment by wcout** [comment]
FWIW: I tried these changes for branch 1.3 and they seem to fix the issue with the test program:
[issue_887_1.3.diff.txt](https://github.com/fltk/fltk/files/13987952/issue_887_1.3.diff.txt)
How did this ever work without releasing the shared image when the filechooser is  closed (gets hidden)?

**Comment by Albrecht-S** [comment]
@wcout Thanks for the patch (not reviewed and not tested by me though).

@MatthiasWM Any idea to fix this in 1.3 as well? ISTR that we discussed releasing 1.3.10 for some other issues anyway and AFAICT we have already two commits since 1.3.9 in branch-1.3.

**Comment by MatthiasWM** [comment]
> @wcout How did this ever work without releasing the shared image when the filechooser is closed (gets hidden)?

Well, it's one bug hiding another bug. By fixing the first one, we revealed the second. My brain is so overexposed to the shared image code that I have a hard time seeing what's wrong and what's right at this point. I will apply your patch to 1.3 and we check from there. 

Then I'll have to go through all the 1.4 changes again. Sorry for the delay.

**Comment by MatthiasWM** [comment]
Ok, I applied the changes for @wcout 's patch and everything seems well in 1.3 . Could you please verify? Thank you.

**Comment by MatthiasWM** [comment]
I checked 1.4 master and I don't get a crash anymore. Did the previous commit https://github.com/fltk/fltk/commit/b2dc8217a17f0e89b303461743965f2b3699a570 fix things for 1.4 ?

**Comment by wcout** [comment]
> @MatthiasWM  I checked 1.4 master and I don't get a crash anymore. Did the previous commit https://github.com/fltk/fltk/commit/b2dc8217a17f0e89b303461743965f2b3699a570 fix things for 1.4 ?

Yes, in addition with commit https://github.com/fltk/fltk/commit/5c90938aa3aa6dcb6f941e8f07937ae1d8951c17.

Branch 1.3 is now also ok.
